src/software/multithreading/thread_safe_buffer.tpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <typeinfo>
+
+#include "software/logger/logger.h"
+
+template <typename T>
+ThreadSafeBuffer<T>::ThreadSafeBuffer(std::size_t buffer_size)
+    : buffer(buffer_size), destructor_called(false)
+{
+}
+
+template <typename T>
+std::optional<T> ThreadSafeBuffer<T>::popLeastRecentlyAddedValue(Duration max_wait_time)
+{
+    // We hold the returned lock in a variable here so that we hold the lock on the
+    // buffer mutex until the lock is destructed at the end of this function
+    auto buffer_lock = waitForBufferToHaveAValue(max_wait_time);
+
+    std::optional<T> result = std::nullopt;
+    if (!buffer.empty())
+    {
+        result = buffer.front();
+        buffer.pop_front();
+    }
+    return result;
+}
+
+template <typename T>
+std::optional<T> ThreadSafeBuffer<T>::popMostRecentlyAddedValue(Duration max_wait_time)
+{
+    // We hold the returned lock in a variable here so that we hold the lock on the
+    // buffer mutex until the lock is destructed at the end of this function
+    auto buffer_lock = waitForBufferToHaveAValue(max_wait_time);
+
+    std::optional<T> result = std::nullopt;
+    if (!buffer.empty())
+    {
+        result = buffer.back();
+        buffer.pop_back();
+    }
+    return result;
+}
+
+template <typename T>
+void ThreadSafeBuffer<T>::push(const T& value)
+{
+    std::scoped_lock<std::mutex> buffer_lock(buffer_mutex);
+    if (buffer.full())
+    {
+        LOG(DEBUG) << "Pushing to a full ThreadSafeBuffer of type: " << typeid(T).name()
```

Comment:
One nit: Do we want this to be `DEBUG` or a higher warning level? `DEBUG` won't show up if we build with `-c opt` (which we would for games). Is this something critical enough we'd want to see it all the time? Maybe just `INFO` so it always shows up but isn't quite a `WARNING`

Reply:
Good point, is it ok for the circular buffer to overwrite sometimes? It seems like a bad idea, so this should be a `WARNING`. Wondering if you can think of a case where we'd intentionally have a full circular buffer

Reply:
Looks like a lot of buffers do overflow, especially ThreadedObservers. I've added an option to suppress the warning

Reply:
Yea I do think there are cases where they will intentionally be full. Eg. anything going to AI will likely be full since the AI is slower than most producers before it.